include/deal.II/matrix_free/operators.h

```diff
@@ -834,6 +884,116 @@ namespace MatrixFreeOperators
       }
   }
 
+
+  //-----------------------------LaplaceOperator----------------------------------
+
+  template <int dim, int fe_degree, int n_components, typename Number>
+  LaplaceOperator<dim, fe_degree, n_components, Number>::
+  LaplaceOperator ()
+    :
+    Base<dim, Number>()
+  {}
+
+
+
+  template <int dim, int fe_degree, int n_components, typename Number>
+  void
+  LaplaceOperator<dim, fe_degree, n_components, Number>::
+  compute_diagonal()
+  {
+    Assert((Base<dim, Number>::data != NULL), ExcNotInitialized());
+
+    unsigned int dummy = 0;
+    LinearAlgebra::distributed::Vector<Number> &inverse_diagonal_entries = Base<dim,Number>::inverse_diagonal_entries;
+    this->initialize_dof_vector(inverse_diagonal_entries);
+    Base<dim,Number>::
+    data->cell_loop (&LaplaceOperator::local_diagonal_cell,
+                     this, inverse_diagonal_entries, dummy);
+
+    const std::vector<unsigned int> &
+    constrained_dofs = Base<dim,Number>::data->get_constrained_dofs();
+    for (unsigned int i=0; i<constrained_dofs.size(); ++i)
+      inverse_diagonal_entries.local_element(constrained_dofs[i]) = 0.;
+    for (unsigned int i=0; i<Base<dim,Number>::edge_constrained_indices.size(); ++i)
+      inverse_diagonal_entries.local_element(Base<dim,Number>::edge_constrained_indices[i]) = 0.;
+
+    for (unsigned int i=0; i<inverse_diagonal_entries.local_size(); ++i)
+      if (std::abs(inverse_diagonal_entries.local_element(i)) > 1e-10)
```

Comment:
Here it would be better to use a stricter tolerance, either `std::sqrt(std::numeric_limits<Number>::epsilon())` or `1e-10*infty_norm` where `infty_norm` is the infinity norm of the vector. For problems with strange scales (I think it is elements of size less than `1e-10` in 3D), the check here would fail for most entries, producing strange results.